src/hotspot/share/opto/mempointer.cpp

```diff
@@ -346,38 +481,144 @@ MemPointerAliasing MemPointerDecomposedForm::get_aliasing_with(const MemPointerD
     return MemPointerAliasing::make_unknown();
   }
 
-  // "MemPointer Lemma" condition (S1):
-  //   Given that all summands are the same, we know that both pointers point into the
-  //   same memory object. With the Pre-Condition, we know that both pointers are in
-  //   bounds of that same memory object.
-
-  // Hence, all 4 conditions of the "MemoryPointer Lemma" are established, and hence
-  // we know that the distance between the underlying pointers is equal to the distance
-  // we computed for the MemPointers:
-  //   p_other - p_this = distance = other.con - this.con
+  if (has_same_base) {
+    // "MemPointer Lemma" condition (S1):
+    //   Given that all summands are the same, we know that both pointers point into the
+    //   same memory object. With the Pre-Condition, we know that both pointers are in
+    //   bounds of that same memory object.
+    //
+    // Hence, all 4 conditions of the "MemPointer Lemma" are established, and hence
+    // we know that the distance between the underlying pointers is equal to the distance
+    // we computed for the MemPointers:
+    //   p_other - p_this = distance = other.con - this.con
 #ifndef PRODUCT
-    if (trace.is_trace_aliasing()) {
-      tty->print_cr("  -> Aliasing always, distance = %d.", distance.value());
-    }
+      if (trace.is_trace_aliasing()) {
+        tty->print_cr("  -> Aliasing always at distance = %d.", distance.value());
+      }
+#endif
+    return MemPointerAliasing::make_always_at_distance(distance.value());
+  } else {
+    // At runtime, the two object bases can be:
+    //   (1) different: pointers do not alias.
+    //   (2) the same:  implies that (S2) holds. The summands are all the same, and with
+    //                  the Pre-Condition, we know that both pointers are in bounds of the
+    //                  same memory object, i.e. (S1) holds. We have already proven (S0)
+    //                  and (S3), so all 4 conditions for "MemPointer Lemma" are given.
+#ifndef PRODUCT
+      if (trace.is_trace_aliasing()) {
+        tty->print_cr("  -> Aliasing not or at distance = %d.", distance.value());
+      }
+#endif
+    return MemPointerAliasing::make_not_or_at_distance(distance.value());
+  }
```

Comment:
You can remove the `else` and just continue because you return in the if-branch:
```suggestion
 
   // At runtime, the two object bases can be:
  //   (1) different: pointers do not alias.
  //   (2) the same:  implies that (S2) holds. The summands are all the same, and with
  //                  the Pre-Condition, we know that both pointers are in bounds of the
  //                  same memory object, i.e. (S1) holds. We have already proven (S0)
  //                  and (S3), so all 4 conditions for "MemPointer Lemma" are given.
#ifndef PRODUCT
  if (trace.is_trace_aliasing()) {
    tty->print_cr("  -> Aliasing not or at distance = %d.", distance.value());
  }
#endif
  return MemPointerAliasing::make_not_or_at_distance(distance.value());
```

Reply:
I have to say I prefer the symmetry here, I would prefer an if-else. Oh but the printing ifs are not correctly aligned.